File: nexal/proxy.py
"""Host-side token proxy — one Unix socket per upstream API.

Each service gets its own socket under /workspace/agents/proxy/:
  - proxy/api.telegram.org  — Telegram Bot API
  - proxy/discord.com       — Discord API

Skill scripts connect to the socket matching the API they need.
The proxy injects the real token and forwards the request.
Tokens never enter the container.
"""

from __future__ import annotations

import json
import logging
import threading
from http.server import BaseHTTPRequestHandler
from pathlib import Path
from socketserver import UnixStreamServer
from typing import Any

import httpx

logger = logging.getLogger("nexal.proxy")

_PROXY_DIR = "proxy"


# ---------------------------------------------------------------------------
# Server
# ---------------------------------------------------------------------------

class _UnixHTTPServer(UnixStreamServer):
    """HTTPServer that listens on a Unix socket."""

    allow_reuse_address = True

    def get_request(self):
        request, client_address = super().get_request()
        return request, ("", 0)


def _start_server(sock_path: Path, handler_cls: type) -> _UnixHTTPServer:
    sock_path.parent.mkdir(parents=True, exist_ok=True)
    sock_path.unlink(missing_ok=True)
    server = _UnixHTTPServer(str(sock_path), handler_cls)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    logger.info("proxy_started socket=%s", sock_path)
    return server


# ---------------------------------------------------------------------------
# Telegram proxy
# ---------------------------------------------------------------------------

_TELEGRAM_UPSTREAM = "https://api.telegram.org"


def _make_telegram_handler(token: str) -> type[BaseHTTPRequestHandler]:

    class TelegramHandler(BaseHTTPRequestHandler):
        def log_message(self, fmt: str, *args: Any) -> None:
            logger.debug(fmt, *args)

        def do_POST(self) -> None:
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length) if content_length else b""
            try:
                data = json.loads(body) if body else {}
            except json.JSONDecodeError:
                return self._respond(400, {"error": "Invalid JSON"})

            # Map path to upstream: POST /sendMessage → api.telegram.org/bot<token>/sendMessage
            method = self.path.strip("/")
            if not method:
                return self._respond(400, {"error": "Missing API method in path"})
            url = f"{_TELEGRAM_UPSTREAM}/bot{token}/{method}"

            try:
                resp = httpx.post(url, json=data, timeout=30)
                self._respond(resp.status_code, resp.json())
            except Exception as e:
                self._respond(502, {"error": str(e)})

        def do_GET(self) -> None:
            method = self.path.strip("/")
            if not method:
                return self._respond(200, {"status": "ok"})
            url = f"{_TELEGRAM_UPSTREAM}/bot{token}/{method}"
            try:
                resp = httpx.get(url, timeout=30)
                self._respond(resp.status_code, resp.json())
            except Exception as e:
                self._respond(502, {"error": str(e)})

        def _respond(self, status: int, data: dict) -> None:
            body = json.dumps(data, ensure_ascii=False).encode()
            self.send_response(status)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)

    return TelegramHandler


# ---------------------------------------------------------------------------
# Discord proxy
# ---------------------------------------------------------------------------

_DISCORD_UPSTREAM = "https://discord.com/api/v10"


def _make_discord_handler(token: str) -> type[BaseHTTPRequestHandler]:

    class DiscordHandler(BaseHTTPRequestHandler):
        def log_message(self, fmt: str, *args: Any) -> None:
            logger.debug(fmt, *args)

        def do_POST(self) -> None:
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length) if content_length else b""
            try:
                data = json.loads(body) if body else {}
            except json.JSONDecodeError:
                return self._respond(400, {"error": "Invalid JSON"})

            path = self.path.rstrip("/")
            if not path:
                return self._respond(400, {"error": "Missing API path"})
            url = f"{_DISCORD_UPSTREAM}{path}"
            headers = {"Authorization": f"Bot {token}"}

            try:
                resp = httpx.post(url, json=data, headers=headers, timeout=30)
                self._respond(resp.status_code, resp.json())
            except Exception as e:
                self._respond(502, {"error": str(e)})

        def do_GET(self) -> None:
            path = self.path.rstrip("/")
            if not path:
                return self._respond(200, {"status": "ok"})
            url = f"{_DISCORD_UPSTREAM}{path}"
            headers = {"Authorization": f"Bot {token}"}
            try:
                resp = httpx.get(url, headers=headers, timeout=30)
                self._respond(resp.status_code, resp.json())
            except Exception as e:
                self._respond(502, {"error": str(e)})

        def _respond(self, status: int, data: dict) -> None:
            body = json.dumps(data, ensure_ascii=False).encode()
            self.send_response(status)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)

    return DiscordHandler


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def _proxy_dir(workspace_dir: str) -> Path:
    return Path(workspace_dir).joinpath("agents", _PROXY_DIR)


def start_proxies(
    workspace_dir: str,
    telegram_token: str | None = None,
    discord_token: str | None = None,
) -> list[_UnixHTTPServer]:
    """Start per-service proxy servers. Returns list of servers for shutdown."""
    base = _proxy_dir(workspace_dir)
    servers: list[_UnixHTTPServer] = []

    if telegram_token:
        handler = _make_telegram_handler(telegram_token)
        servers.append(_start_server(base / "api.telegram.org", handler))

    if discord_token:
        handler = _make_discord_handler(discord_token)
        servers.append(_start_server(base / "discord.com", handler))

    return servers

File: crates/agent/src/proxy.rs
//! Host-side token proxy — Unix sockets for API access without exposing tokens.
//!
//! Each service gets its own socket under `<workspace>/agents/proxy/`:
//!   - `proxy/api.telegram.org` — Telegram Bot API
//!   - `proxy/discord.com` — Discord API
//!
//! Skill scripts inside the container connect to the socket.
//! The proxy injects the real token and forwards to the upstream API.
//! Tokens never enter the container.

use std::path::{Path, PathBuf};
use std::sync::Arc;

use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt, BufReader};
use tokio::net::UnixListener;
use tracing::{debug, error, info, warn};

/// Start all configured proxy servers. Returns handles to stop them.
pub async fn start_proxies(
    workspace: &Path,
    telegram_token: Option<&str>,
    discord_token: Option<&str>,
) -> Vec<tokio::task::JoinHandle<()>> {
    let proxy_dir = workspace.join("agents").join("proxy");
    let _ = tokio::fs::create_dir_all(&proxy_dir).await;

    let mut handles = Vec::new();

    if let Some(token) = telegram_token {
        let sock = proxy_dir.join("api.telegram.org");
        let handle = start_http_proxy(
            sock,
            "https://api.telegram.org".to_string(),
            format!("bot{token}"),
        )
        .await;
        handles.push(handle);
    }

    if let Some(token) = discord_token {
        let sock = proxy_dir.join("discord.com");
        let handle = start_http_proxy(
            sock,
            "https://discord.com".to_string(),
            format!("Bot {token}"),
        )
        .await;
        handles.push(handle);
    }

    handles
}

/// Start a single HTTP proxy on a Unix socket.
///
/// Accepts simple HTTP/1.1 POST requests, injects auth, forwards upstream.
async fn start_http_proxy(
    sock_path: PathBuf,
    upstream_base: String,
    auth_value: String,
) -> tokio::task::JoinHandle<()> {
    // Remove stale socket
    let _ = tokio::fs::remove_file(&sock_path).await;

    let listener = match UnixListener::bind(&sock_path) {
        Ok(l) => l,
        Err(e) => {
            error!(path = %sock_path.display(), "failed to bind proxy socket: {e}");
            return tokio::spawn(async {});
        }
    };

    // Make socket world-writable so container (different uid) can connect
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let _ = std::fs::set_permissions(&sock_path, std::fs::Permissions::from_mode(0o777));
    }

    info!(path = %sock_path.display(), upstream = %upstream_base, "proxy started");

    let upstream_base = Arc::new(upstream_base);
    let auth_value = Arc::new(auth_value);

    tokio::spawn(async move {
        let client = reqwest::Client::new();

        loop {
            let (stream, _) = match listener.accept().await {
                Ok(s) => s,
                Err(e) => {
                    warn!("proxy accept error: {e}");
                    continue;
                }
            };

            let client = client.clone();
            let upstream = Arc::clone(&upstream_base);
            let auth = Arc::clone(&auth_value);

            tokio::spawn(async move {
                if let Err(e) = handle_connection(stream, &client, &upstream, &auth).await {
                    debug!("proxy connection error: {e}");
                }
            });
        }
    })
}

/// Handle a single HTTP request over a Unix socket connection.
///
/// Parses a minimal HTTP/1.1 POST, forwards to upstream with auth, returns response.
async fn handle_connection(
    stream: tokio::net::UnixStream,
    client: &reqwest::Client,
    upstream_base: &str,
    auth_value: &str,
) -> anyhow::Result<()> {
    let (reader, mut writer) = stream.into_split();
    let mut reader = BufReader::new(reader);

    // Read request line: POST /path HTTP/1.1
    let mut request_line = String::new();
    reader.read_line(&mut request_line).await?;
    let parts: Vec<&str> = request_line.trim().split_whitespace().collect();
    if parts.len() < 2 {
        writer.write_all(b"HTTP/1.1 400 Bad Request\r\n\r\n").await?;
        return Ok(());
    }
    let method = parts[0];
    let path = parts[1];

    // Read headers
    let mut content_length: usize = 0;
    loop {
        let mut line = String::new();
        reader.read_line(&mut line).await?;
        if line.trim().is_empty() {
            break;
        }
        if let Some(val) = line.strip_prefix("Content-Length:").or_else(|| line.strip_prefix("content-length:")) {
            content_length = val.trim().parse().unwrap_or(0);
        }
    }

    // Read body
    let mut body = vec![0u8; content_length];
    if content_length > 0 {
        reader.read_exact(&mut body).await?;
    }

    // Forward to upstream
    let url = format!("{upstream_base}/{auth_value}/{}", path.trim_start_matches('/'));

    let upstream_resp = match method {
        "POST" => {
            client
                .post(&url)
                .header("content-type", "application/json")
                .body(body)
                .send()
                .await
        }
        "GET" => client.get(&url).send().await,
        _ => {
            writer
                .write_all(b"HTTP/1.1 405 Method Not Allowed\r\n\r\n")
                .await?;
            return Ok(());
        }
    };

    let resp = match upstream_resp {
        Ok(r) => r,
        Err(e) => {
            let msg = format!("HTTP/1.1 502 Bad Gateway\r\n\r\n{{\"error\":\"{e}\"}}");
            writer.write_all(msg.as_bytes()).await?;
            return Ok(());
        }
    };

    let status = resp.status();
    let resp_body = resp.bytes().await.unwrap_or_default();

    let response = format!(
        "HTTP/1.1 {status}\r\nContent-Type: application/json\r\nContent-Length: {}\r\n\r\n",
        resp_body.len()
    );
    writer.write_all(response.as_bytes()).await?;
    writer.write_all(&resp_body).await?;

    Ok(())
}
